File: Rust/Performance/bounded_latency/bounded_latency_multi_thread.rs
/*
===============================================================================
 Project    : Assessing Python’s Suitability for Airborne Safety-Critical
                       Systems under DO-178C Guidelines
 File       : bounded_latency_multi_thread.rs
 Author(s)  : Mohammad Mari, Lian Wen
 Affiliation: School of ICT, Griffith University
 Contact    : mohammad.mari@griffithuni.edu.au
 Created    : 2025
 License    : MIT License (see LICENSE file for details)
 Description: Core experiment code to demonstrate bounded_latency in rust
 Usage      : rustc -C opt-level=0 bounded_latency_multi_thread.rs
===============================================================================
*/

// This example runs multiple simulated sensor processing tasks in parallel 
// using Rust threads. Unlike Python, Rust has no Global Interpreter Lock (GIL),
// so CPU-bound tasks can execute concurrently across cores. The results can be
// compared with the single-threaded and Python versions to highlight differences
// in latency.
// Output:
//   sens4 finished in 0.05s, expected <= 0.10s, delay -0.05s
//   sens2 finished in 0.06s, expected <= 0.10s, delay -0.04s
//   sens1 finished in 0.06s, expected <= 0.10s, delay -0.04s
//   sens3 finished in 0.06s, expected <= 0.10s, delay -0.04s

use std::thread;
use std::time::{Instant};

const NUM_SENSORS: usize = 4;
const EXPECTED_FINISH_TIME_SEC: f64 = 0.1;

// Simulate sensor processing functionality
fn process(sensor_id: String, start_time: Instant, expected_finish_time_sec: f64) {
    let mut some_num: u64 = 0;

    for i in 0..10_000_000 {
        some_num = some_num.wrapping_add(i);
    }

    let end_time = Instant::now();
    let actual_time = end_time.duration_since(start_time).as_secs_f64();
    let delay = actual_time - expected_finish_time_sec;
    println!(
        "{} finished in {:.2}s, expected <= {:.2}s, delay {:.2}s",
        sensor_id, actual_time, expected_finish_time_sec, delay
    );
}

fn main() {
    let start_time = Instant::now();
    let mut handles = vec![];

    // Create and start all threads
    for i in 1..=NUM_SENSORS {
        let sensor_id = format!("sens{}", i);
        let start_time_clone = start_time.clone();

        let handle = thread::spawn(move || {
            process(sensor_id, start_time_clone, EXPECTED_FINISH_TIME_SEC);
        });

        handles.push(handle);
    }

    // Wait for all threads to finish
    for handle in handles {
        handle.join().unwrap();
    }
}

File: Python/Performance/bounded_latency/bounded_latency_multi_thread.py
"""
===============================================================================
 Project    : Assessing Python’s Suitability for Airborne Safety-Critical
                       Systems under DO-178C Guidelines
 File       : bounded_latency_multi_thread.py
 Author(s)  : Mohammad Mari, Lian Wen
 Affiliation: School of ICT, Griffith University
 Contact    : mohammad.mari@griffithuni.edu.au
 Created    : 2025
 License    : MIT License (see LICENSE file for details)
 Description: Core experiment code to demonstrate bounded_latency and GIL impact
 Usage      : python bounded_latency_multi_thread.py
              for profiling: 1. mprof run python bounded_latency_multi_thread.py
                             2. mprof plot   
===============================================================================
"""

# This demo illustrates the impact of the Global Interpreter Lock (GIL) 
# on CPU-bound tasks when using multiple threads in Python. 

# Due to the GIL, only one thread executes Python bytecode at a time, 
# which limits the potential speedup from multi-threading. 
# This effect is especially visible in CPU-bound tasks, 
# where threads are forced to run sequentially rather than in parallel.
# In this example all sensors failed.
# Output: 
#   sens1 finished in 1.11s, expected <= 1s, delay 0.11s
#   sens2 finished in 1.12s, expected <= 1s, delay 0.12s
#   sens3 finished in 1.15s, expected <= 1s, delay 0.15s
#   sens4 finished in 1.16s, expected <= 1s, delay 0.16s


import threading
import time


NUM_SENSORS = 4
EXPECTED_FINISH_TIME_SEC = 1

# Simulate sensor processing functionality
def process(sensor_id, start_time, expected_finish_time_sec):

    some_num = 0

    for i in range(10_000_000):
        some_num += i
    end_time = time.time()
    actual_time = end_time - start_time
    delay = actual_time - expected_finish_time_sec
    print(
        f"{sensor_id} finished in {actual_time:.2f}s, expected <= {expected_finish_time_sec}s, delay {delay:.2f}s"
    )


threads = []
start_time = time.time()

# Create and start all threads
for i in range(1, NUM_SENSORS + 1):
    t = threading.Thread(target=process, args=(f"sens{i}", start_time, EXPECTED_FINISH_TIME_SEC))
    threads.append(t)
    t.start()

# Wait for all threads to finish
for t in threads:
    t.join()